### PR TITLE
Created a product webservice that queries based on source, type, code…

### LINKED ISSUE #281: Product Detail Service
required query parameters
- source
- type
- code

optional query parameters
- updateTime
    ISO8601 format
    when specified, find specific version of product.
    when not specified, find most recent version of product.
- path
    when path is specified, find product as usual.
    If content path does not exist, 404 Not Found.
    If content path is the empty string (""), respond with content (content type and bytes)
    Otherwise, if updateTime was specified respond with 302 redirect with content url in Location header

output format
- geojson, similar to individual products in fdsn geojson detail output format
see https://github.com/usgs/earthquake-event-ws/blob/master/src/lib/classes/pdl/Event.class.php#L627

- set last modified header to product updateTime

caching
- when updateTime is specified as a query parameter, use a far-future caching header (1 year)
- when updateTime is not specified, and finding most recent version, use a 1 minute caching header.  Also check for request headers for "If-Modified-Since", respond with "304 Not Modified" if product has not been updated since that time.

The ProductIndex class has a method to get a ProductSummary object, once its index id is known, but does not have any methods to search based on source, type, code, and updateTime.
https://github.com/usgs/earthquake-event-ws/blob/master/src/lib/classes/pdl/ProductIndex.class.php

Use similar error formatting to FDSN service.
see https://github.com/usgs/earthquake-event-ws/blob/master/src/lib/classes/fdsn/FDSNEventWebService.class.php#L327

Consider creating a ProductWebService class (and parent web service class if that helps avoid duplication...).

### PR DESCRIPTION
**Pull request by Capprin**
…, and optional updateTime. Does not currently include path searching.

closes #281

### CODE REVIEW

File: src/lib/classes/fdsn/FDSNEventWebService.class.php [hunk 1]
```diff
@@ -1,35 +1,12 @@
 <?php
 
+class FDSNEventWebService extends WebService{
```
Comment (jmfee-usgs) [inline src/lib/classes/fdsn/FDSNEventWebService.class.php]: Add a space before opening curly braces.

File: src/lib/classes/pdl/ProductIndex.class.php [hunk 2]
```diff
@@ -1605,5 +1607,60 @@ public function getConnection() {
     return $this->connection;
   }
 
+  //Searches for product based on source, type, and code
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductIndex.class.php]: Put these comments as the first line inside the /* */ comment block below.

File: src/lib/classes/pdl/ProductIndex.class.php [hunk 3]
```diff
@@ -1605,5 +1607,60 @@ public function getConnection() {
...
+
+    //Construct prepared WHERE statement from key query terms
+    $where = sprintf(
+      "%s=? AND %s=? AND %s=?",
+      self::SUMMARY_TYPE,
+      self::SUMMARY_SOURCE,
+      self::SUMMARY_CODE
+    );
+    //initializing prepared statement
+    $params = array($query->type,$query->source,$query->code);
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductIndex.class.php]: Use spaces after commas.

File: src/lib/classes/pdl/ProductIndex.class.php [hunk 4]
```diff
@@ -1605,5 +1607,60 @@ public function getConnection() {
...
+      $params[] = $query->updateTime; //Update prepared statement
+    }
+
+    //Build SQL, changing table based on existence of updateTime
+    $sql = sprintf('
+      SELECT %s
+      FROM %s
+      WHERE %s',
+      self::SUMMARY_PRODUCT_ID,
+      isset($query->updateTime)?self::SUMMARY_TABLE:self::SUMMARY_CURRENT_TABLE, //Ternary operator to choose between tables
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductIndex.class.php]: Put spaces around ternary operators, e.g. `bool ? true : false`

or if values are long you can wrap and indent:
```
bool
  ? true
  : false
```

File: src/lib/classes/pdl/ProductIndex.class.php [hunk 5]
```diff
@@ -1605,5 +1607,60 @@ public function getConnection() {
...
+      $where
+    );
+
+    //Request information from database using prepared statement
+    $sql = $connection->prepare($sql); 
+    $sql->execute($params); 
+
+    $productId = $sql->fetch()[self::SUMMARY_PRODUCT_ID];
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductIndex.class.php]: Do you need to check for errors, or handle a case if fetch doesn't return a matching value?

File: src/lib/classes/pdl/ProductIndex.class.php [hunk 6]
```diff
@@ -1605,5 +1607,60 @@ public function getConnection() {
     return $this->connection;
   }
 
+  //Searches for product based on source, type, and code
+  /**
+   * @param $query {ProductQuery}
+   *    Productquery storing supplied information
+   */
+  public function getProductByQuery($query){ 
+    global $storage;
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductIndex.class.php]: Update this method to return the productId, and let the service use the returned id to access product storage.

File: src/lib/classes/pdl/ProductWebService.class.php [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  //Constructs query object based on parameters in $_GET
+  protected function parseQuery(){
+    $query = new ProductQuery();
+
+    //Parse fields
+    foreach ($_GET as $name=>$value){
+      if ($value === ''){
+        $this->error(self::BAD_REQUEST, self::$statusMessage[self::BAD_REQUEST]);
+      }
+      elseif ($name == 'source'){
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductWebService.class.php]: Put these on the same line as the previous curly brace.

File: src/lib/classes/pdl/ProductWebService.class.php [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        $query->type = $value;
+      }
+      elseif ($name == 'code'){
+        $query->code = $value;
+      }
+      elseif ($name == 'updatetime' || $name == 'updateTime'){
+        $query->updateTime = $value;
+      }
+      else{
+        $this->error(0,"Invalid field {" . $name . "} supplied.");
```
Comment (jmfee-usgs) [inline src/lib/classes/pdl/ProductWebService.class.php]: escape this output.

File: src/lib/classes/pdl/ProductIndex.class.php (hunk 5 above)
Comment (Capprin) [reply src/lib/classes/pdl/ProductIndex.class.php]: Error handling is done by following conditional statement. Fetch will just return null if we get no response.

File: src/lib/classes/pdl/ProductIndex.class.php (hunk 6 above)
Comment (jmfee-usgs) [reply src/lib/classes/pdl/ProductIndex.class.php]: You may want to update this method to return a ProductSummary object, looking up the `id` and then call and return `getSummary`

**Comment by jmfee-usgs** [comment]
included in #285

**Pull request closed without merging**